text/3703-wrapping-from.md

```diff
@@ -0,0 +1,277 @@
+- Feature Name: `num_wrapping_from`
+- Start Date: 2024-09-13
+- RFC PR: [rust-lang/rfcs#3703](https://github.com/rust-lang/rfcs/pull/3703)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a `core::num::WrappingFrom<T>` trait for converting between integer types which might truncate.
+
+
+# Motivation
+[motivation]: #motivation
+
+Today you can convert between arbitrary primitive integer types with `as`.
+
+But there's a few ways in which it'd be nice to have something else:
+
+1. You can't use it in generic code.
+2. You can't use it with non-primitive types like `BigInteger`.
+3. `as` can do enough other things -- like the [evil] `i32::max as u32` -- that you might want to
+   more clearly communicate the intent by using a more specific API.
+
+[evil]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bddbeb692d342b8ebfefa1d25be9275d
+
+This PR thus proposes a new `WrappingFrom` trait for these scenarios:
+
+1. You can `where T: WrappingFrom<U>` in your generics.
+2. The crate defining `BigInteger` can `impl WrappingFrom<BigInteger> for u32 { … }` and such.
+3. You can write `u32::wrapping_from(x)` for a conversion if you prefer, and `u32::from(i32::max)` just won't compile.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We've seen the `From` trait already, which is for *lossless* and *value-preserving* conversions.
+With integers, though, you might sometimes want *truncating* conversions that you can't do via `From`.
+For that, you can use the `num::WrappingFrom` trait instead.
+
+`WrappingFrom` is specific to *numbers*.  You can't use it for things like `String: From<&str>` which we've used a bunch.
+
+It's is implemented for the numeric widening conversions that also exist in `From`.
+So `i128: From<i32>` and `i128: WrappingFrom<i32>` both exist.
+
+But while `i32: From<i128>` intentionally doesn't exist -- as it'd be lossy -- there *is*
+an `i32: WrappingFrom<i128>` implementation which truncates to just give the bottom 32 bits.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+// in core::num (and std::num)
+
+/// Performs potentially-lossy conversions in a quantized numeric lattice.
```

Comment:
What is a "quantized numeric lattice"?

Could this be something like "Performs potentially-lossy conversion for integral types"? Does it make sense for non-integer types? For floating points, I'm not sure when you would ever want it to wrap instead of produce an infinite value if the original is out of range. 

This also doesn't really describe in what way the conversion is lossy, although that is discussed elsewhere in the RFC.  It would probably be good to have simple description (it truncates), and a more formal description of the semantics, perhaps something like:

>  For an integer `a: Big` and a smaller integer type `Small`, `Small::wrapping_from(a)` returns a value `b` in `Small` such that `a == b + (R * n)` for some `n` in `Big` where `R` is the number of possible values in `Small` ( that is effectively `Small::MAX - Small::MIN + 1` ).

Reply:
> returns a value `b` in `Small` such that `a == b + (R * n)` for some `n` in `Big` where `R` is the number of possible

I would phrase that:

returns the unique value $b$ in `Small` such that, in unlimited range mathematical integers, $a = b + n \times R$ where $R$ is the number of possible

note I'm using latex math notation `$a = b + n \times R$`